src/Domain/LeanCode.DomainModels.MassTransitRelay/Testing/AsyncManualResetEvent.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using GreenPipes.Internals.Extensions;
+
+namespace LeanCode.DomainModels.MassTransitRelay.Testing
+{
+    // Based on https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Coordination/AsyncManualResetEvent.cs
+    public sealed class AsyncManualResetEvent
+    {
+        private readonly object mutex;
+        private TaskCompletionSource<object?> tcs;
```

Comment:
I'd prefer `TaskCompletionSource<ValueTuple>` over `TaskCompletionSource<object?>` if you don't use it to store an actual result, just signal completion of the task.

Reply:
Using `object?` with `null` as a marker value is shorter and, at least now, more prevalent in .NET world. Also, it results in shorter code since `()` is not an empty tuple and you have to use `new ValueTuple()` (which, I think, will be boxed in the end, whereas `null` won't).

Reply:
You could use `TrySetResult(default)`. And boxing shouldn't be an issue, unless `TaskCompletionSource` stores it internally as a reference anyway. Actually, I'm going to check this later, but for now `object?` is fine.

Reply:
Okay, they [have one with `VoidResult`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Threading.Channels/src/System/Threading/Channels/TaskCompletionSource.cs) in Channels, so I think even .NET team is going away from `object?` here.

Reply:
And no, they don't store it directly, but only as part of `Task<T>`, so there won't be any additional allocation.